compiler/vm/vmdef.nim

```diff
@@ -90,27 +91,224 @@ type
+    # XXX `setElem` could also be run-length encoded for speed and memory efficiency...
+
+    # XXX In the end, VmSets should be embedded into their surrounding storage, without any heap allocations
+    elems* : seq[uint16]
+
+  Atom* {.union.} = object
+    intVal*: BiggestInt ## akInt
+    floatVal*: BiggestFloat ## akFloat
+    ptrVal*: ptr Atom ## akPtr
+    setVal*: VmSet ## akSet
+    strVal*: VmString ## akString
+    seqVal*: VmSeq ## akSeq
+    refVal*: GlobalRef ## akRef
+    callableVal*: VmFunctionObject ## akCallable
+
+    nodeVal*: PNode ## akPNode
+
+  # TODO: needs a better name
+  VmObjectAtom* = object ## A container that stores a value. Single atom
+                         ## values are stored in-place while complex values
+                         ## are seperately allocated 
+    case kind*: AtomKind
+    of akInt..pred(akObject):
+      atom: array[32, byte]
+    of akObject, akArray:
+      obj*: MemoryHandle
+    
+    typ*: PVmType
+
+  
+  # XXX: VmObjectAtom doesn't store pseudo atoms in-place, due to their dynamic size.
+  # However, `TypedVmObject`s are only meant to be used for globals and constants. For these, 
+  # the pseudo-atom flat byte size is both constant and known at initialization time. Thus, 
+  # we could allocate the raw atom or object (pseudo atom) in the same memory location. We'd need a custom
+  # allocator for `TypedVmValue` then, but globals and constants probably need one anyway...
+  TypedVmValue* = VmObjectAtom
+
+  TFullReg* = object
     case kind*: TRegisterKind
-    of rkNone: nil
+    of rkNone: discard
     of rkInt: intVal*: BiggestInt
     of rkFloat: floatVal*: BiggestFloat
-    of rkNode: node*: PNode
-    of rkRegisterAddr: regAddr*: ptr TFullReg
-    of rkNodeAddr: nodeAddr*: ptr PNode
-
+    of rkAddress:
+      addrVal*: MemoryHandle
+      addrTyp*: PVmType
+      isValid*: bool ## Indicates if `addrVal` points to a valid location
+    of rkHandle, rkValue:
+      handle*: TypedHandle
+      global*: Option[GlobalRef] ## If valid, the global of which the handle references a location
+    of rkNimNode:
+      nimNode*: PNode
+    
+    of rkRegAddr:
+      regAddr*: ptr TFullReg
+    
+
+  # XXX should probably be a distinct int
+  GlobalRef* = int ## A reference to an entry in `TGlobals`
+
+  Global = ref object
+    # TODO: would uint32 be enough for refCount?
+    o: TypedVmValue
+    refCount: int
+    # XXX: The PType is currently still needed for a few things:
+    # - type relation
+    # - lifetime hooks
+    # It should be not that hard to work around having to use PType for these
+
+    nType: PType
+
+
+  TGlobals* = object
+    # TODO: use a different memory management strategy for keeping references to globals valid
+    # Right now I just use refs (we don't need/use the shared ownership semantics, just the automatic garbage collection)
+    # to keep globals alive when `data` changes it's length
```

Comment:
I'm a bit confused, if you only refer to globals via their index, then why would you need `ref` at all?

Reply:
The way global/heap objects are handled here is already outdated/obsolete. Regarding the version you're looking at, consider this case:
```
let a = new int
let b = addr(a[])
let c = new int
b[] = 1
```
Small values (e.g. `int`), were stored directly inside the `VmObjectAtom` (which is what a global/heap object was stored as). If globals would have been stored directly inside the `globals` seq, then their actual address in memory would have changed whenever the `seq`s internal storage was resized. Using `new` (`opcNew`) adds a new item to `globals`, thus maybe triggering a resize.

In the above code example, this would have then lead to `b[] = 1` segfaulting since `b` would have been a pointer into a now freed memory location. In order to not having to do manual memory management via `alloc`/`dealloc` back when I implemented this, I simply used `ref` for globals/heap objects.

But again, the heap object management you're seeing here is already outdated.